### ISSUE TITLE
Read/Write Custom Png Chunk

### ISSUE THREAD

**Issue by boban984** [JimBobSquarePants/ImageSharp#251]
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of ImageSharp
- [ ] I have verified if the problem exist in both `DEBUG` and `RELEASE` mode
- [x] I have searched [open](https://github.com/JimBobSquarePants/ImageSharp/issuess) and [closed](https://github.com/JimBobSquarePants/ImageSharp/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
I have a classic .net app where i read/write custom png chunk information (using the Pngcs lib), but now i need to do the same on a .net core application.
I saw the source for "PngDecoderCore.cs" and it has a "ReadChunk" method, when i load a Png with "tEXt" chunks it succesufully reads them.

Is it possible to create a WriteChunk method and also a ReadChunk method witch takes an "Id" parameter for the chunk.

Maybe i can help out with the "WriteChunk" method.

Thanks

### System Configuration
Windows 10 64bit

- ImageSharp version: 1.0.0-alpha9-00140
- Environment (Operating system, version and so on):
- .NET Framework version: .net core 1.1
- Additional information:

**Comment by JimBobSquarePants** [comment]
Hi @boban984 

Yeah I guess it should be possible, we'd probably have to pass the chunks as extra params to the encoder.If you'd fancy lending a hand that would be great! 👍 

Cheers

James

**Comment by dlemstra** [comment]
I think we should add that extra chunk to the `IPngEncoderOptions` to write it to the file. And when it is read it should be added to the `MetaData` of the `Image`. Maybe use an extension method?

**Comment by boban984** [comment]
Great that you are open to the idea, i will start implementing the read/write of the custom chunk.
I will add it to the Encoder options for writing and then make it part of the `MetaData`, i guess this shouldn't take me more than a week.

If you guys have any advice or things to look out for before, pls tell me. Thanks

**Comment by dlemstra** [comment]
@boban984 It might be wise to wait till made those changes :) The public API is very likely to change.

**Comment by vpenades** [comment]
As an API suggestion:

Custom PNG chunks can be very large, I've seen some applications that use a custom chunk to store the actual data in a PNG container that uses the PNG image as the thunbnail of the actual content.

So, since these chunks can only be useful for those that specifically want to load them, I would suggest to set a programmable filter in PNGDecoder, so only the chunk tags that are in the filter are copied to the MetaData structure.

In fact, consider PNG as a general container with a preview thumbnail.

**Comment by JimBobSquarePants** [comment]
I'll have to have a good read of the spec again to determine what is what.

**Comment by JD1337** [comment]
> Great that you are open to the idea, i will start implementing the read/write of the custom chunk.
> I will add it to the Encoder options for writing and then make it part of the `MetaData`, i guess this shouldn't take me more than a week.
> 
> If you guys have any advice or things to look out for before, pls tell me. Thanks

Did you end up making any progress for this? Was looking for this exact thing now and came across this.

**Comment by JimBobSquarePants** [comment]
@JD1337 No, it's marked as a future milestone as it's low priority and we have limited resources. Happy to review and accept any potential PRs though.

**Comment by JimBobSquarePants** [comment]
I want to return to this issue and check what is the actual requirement here?

ImageSharp writes custom data as text-type chunks, choosing the correct one depending on the length. 

https://github.com/SixLabors/ImageSharp/blob/da6d56bc62ed90c26a4c37a7dbba55f6d90ddae1/src/ImageSharp/Formats/Png/PngEncoderCore.cs#L1111

I think that actually closes the requirements of the open issue.
